applications/vanilla/models/class.commentmodel.php

```diff
@@ -654,118 +654,118 @@ public function getUserScore($commentID, $userID) {
-                            }
-
-                            // Mark this category read if all the new content is read
-                            if ($markAsRead) {
-                                $categoryModel = new CategoryModel();
-                                $categoryModel->saveUserTree($categoryID, ['DateMarkedRead' => Gdn_Format::toDateTime()]);
-                                unset($categoryModel);
-                            }
-
-                        }
-                    }
-                }
+        } else {
+            // Make sure the discussion isn't archived.
+            $archiveDate = c('Vanilla.Archive.Date', false);
+            if (!$archiveDate || (Gdn_Format::toTimestamp($discussion->DateLastComment) > Gdn_Format::toTimestamp($archiveDate))) {
+                $newComments = true;
+
+                // Insert watch data.
+                $this->SQL->options('Ignore', true);
+                $this->SQL->insert(
+                    'UserDiscussion',
+                    [
+                        'UserID' => $userID,
+                        'DiscussionID' => $discussion->DiscussionID,
+                        'CountComments' => $countWatch,
+                        'DateLastViewed' => Gdn_Format::toDateTime()
+                    ]
+                );
             }
+        }
 
+        /**
+         * Fuzzy way of trying to automatically mark a category read again
+         * if the user reads all the comments on the first few pages.
+         */
+
+        // If this discussion is in a category that has been marked read,
+        // check if reading this thread causes it to be completely read again.
+        $categoryID = val('CategoryID', $discussion);
+        if (!$categoryID) {
+            return;
+        }
+        $category = CategoryModel::categories($categoryID);
+        if (!$category) {
+            return;
+        }
+        $dateMarkedRead = val('DateMarkedRead', $category);
+        if (!$dateMarkedRead) {
+            return;
         }
+        // Fuzzy way of looking back about 2 pages into the past.
+        $lookBackCount = c('Vanilla.Discussions.PerPage', 50) * 2;
+
+        // Find all discussions with content from after DateMarkedRead.
+        $discussionModel = new DiscussionModel();
+        $discussions = $discussionModel->get(0, $lookBackCount + 1, [
+            'CategoryID' => $categoryID,
+            'DateLastComment>' => $dateMarkedRead
+        ]);
+        unset($discussionModel);
+
+        // Abort if we get back as many as we asked for, meaning a
+        // lot has happened.
+        if ($discussions->numRows() > $lookBackCount) {
+            return;
+        }
+
+        // Loop over these discussions and see if any are still unread.
+        $markAsRead = true;
+        while ($discussion = $discussions->nextRow(DATASET_TYPE_ARRAY)) {
+            if ($discussion['Read']) {
+                continue;
+            }
+            $markAsRead = false;
```

Comment:
I think we can still shorten this anyways. As is you aren't actually checking for the value of `$markAsRead` anymore. I'm going try a different type of loop here.


```php
// Bail out if there are any unread discussions
while ($discussion = $discussions->nextRow(DATASET_TYPE_ARRAY)) {
    if (!$discussion['Read']) {
        return
    }
}
```